ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+        }
+    }
+
+  private[this] val completeOnLedgerEnd = LedgerTest(
+    "TXCompleteOnLedgerEnd",
+    "A stream should complete as soon as the ledger end is hit") { ledger =>
+    val transactionsToSubmit = 14
+    for {
+      party <- ledger.allocateParty()
+      transactionsFuture = ledger.flatTransactions(party)
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+        contract <- ledger.create(party, Dummy(party))
+      } yield contract))
+      _ <- transactionsFuture
+    } yield {
+      // doing nothing: we are just checking that `transactionsFuture` completes successfully
+    }
+  }
+
+  private[this] val processInTwoChunks = LedgerTest(
+    "TXProcessInTwoChunks",
+    "Serve the complete sequence of transactions even if processing is stopped and resumed") {
+    ledger =>
+      val transactionsToSubmit = 5
+      for {
+        party <- ledger.allocateParty()
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterFirstSection <- ledger.currentEnd()
+        firstSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.end := endAfterFirstSection)
+        firstSection <- ledger.flatTransactions(firstSectionRequest)
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterSecondSection <- ledger.currentEnd()
+        secondSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.begin := endAfterFirstSection, _.end := endAfterSecondSection)
+        secondSection <- ledger.flatTransactions(secondSectionRequest)
+        fullSequence <- ledger.flatTransactions(party)
+      } yield {
+        val concatenation = Vector.concat(firstSection, secondSection)
+        assert(
+          fullSequence == concatenation,
+          s"The result of processing items in two chunk should yield the same result as getting the overall stream of transactions in the end but there are differences. " +
+            s"Full sequence: ${fullSequence.map(_.commandId).mkString(", ")}, " +
+            s"first section: ${firstSection.map(_.commandId).mkString(", ")}, " +
+            s"second section: ${secondSection.map(_.commandId).mkString(", ")}"
+        )
+      }
+  }
+
+  private[this] val identicalAndParallel = LedgerTest(
+    "TXParallel",
+    "The same data should be served for two identical, parallel requests") { ledger =>
+    val transactionsToSubmit = 5
+    val parallelRequests = 10
+    for {
+      party <- ledger.allocateParty()
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+        contract <- ledger.create(party, Dummy(party))
+      } yield contract))
+      results <- Future.sequence(Vector.fill(parallelRequests)(for {
```

Comment:
Same remark as above regarding the for-comprehension.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2713/commits/d795e4186ce1618edf2695cb40f6148c0e3f896d